core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/common/index/JoinerUtils.java

```diff
@@ -37,92 +37,147 @@ private JoinerUtils() {
 
     public static <A, B> Function<A, IndexProperties> combineLeftMappings(DefaultBiJoiner<A, B> joiner) {
         int joinerCount = joiner.getJoinerCount();
-        if (joinerCount == 0) {
-            return (A a) -> NONE_INDEX_PROPERTY;
-        } else if (joinerCount == 1) {
-            Function<A, Object> mapping = joiner.getLeftMapping(0);
-            return (A a) -> new SingleIndexProperties(mapping.apply(a));
-        } else {
-            return (A a) -> {
-                Object[] mappings = new Object[joinerCount];
-                for (int i = 0; i < joinerCount; i++) {
-                    mappings[i] = joiner.getLeftMapping(i).apply(a);
-                }
-                return new ManyIndexProperties(mappings);
-            };
+        switch (joinerCount) {
+            case 0:
+                return (A a) -> NONE_INDEX_PROPERTY;
+            case 1:
+                Function<A, Object> mapping = joiner.getLeftMapping(0);
+                return (A a) -> new SingleIndexProperties(mapping.apply(a));
+            case 2:
+                Function<A, Object> mappingX = joiner.getLeftMapping(0);
+                Function<A, Object> mappingY = joiner.getLeftMapping(1);
+                return (A a) -> new TwoIndexProperties(mappingX.apply(a), mappingY.apply(a));
+            case 3:
+                Function<A, Object> mapping1 = joiner.getLeftMapping(0);
+                Function<A, Object> mapping2 = joiner.getLeftMapping(1);
+                Function<A, Object> mapping3 = joiner.getLeftMapping(2);
+                return (A a) -> new ThreeIndexProperties(mapping1.apply(a), mapping2.apply(a), mapping3.apply(a));
```

Comment:
The naming of mapping variables is a bit inconsistent (x,y vs. 1,2,3). See the other methods of this class too.

Reply:
It is on purpose. Different branches of a switch statement are considered the same scope, so I can not reuse variable names. Dumb, but that's just Java.

Reply:
Good to know, I was not aware of that.